test/Common.FunctionalTests/AppOfflineTests.cs

```diff
@@ -29,32 +30,114 @@ public class AppOfflineTests : IISFunctionalTestBase
             AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
 
             await AssertAppOffline(deploymentResult);
+            AssertFilesNotLocked(deploymentResult);
         }
 
-        [ConditionalTheory]
-        [InlineData(HostingModel.InProcess)]
-        [InlineData(HostingModel.OutOfProcess)]
-        public async Task AppOfflineDroppedWhileSiteIsDown_CustomResponse(HostingModel hostingModel)
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteFailedToStartInShim_AppOfflineServed_InProcess()
         {
-            var expectedResponse = "The app is offline.";
-            var deploymentResult = await DeployApp(hostingModel);
+            var deploymentResult = await DeployApp(HostingModel.InProcess);
+
+            Helpers.ModifyAspNetCoreSectionInWebConfig(deploymentResult, "processPath", "nonexistent");
+
+            var result = await deploymentResult.HttpClient.GetAsync("/");
+            Assert.Equal(500, (int)result.StatusCode);
+            Assert.Contains("500.0", await result.Content.ReadAsStringAsync());
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            await AssertAppOffline(deploymentResult);
+            AssertFilesNotLocked(deploymentResult);
+        }
+
+        [ConditionalFact]
+        [RequiresIIS(IISCapability.ShutdownToken)]
+        public async Task AppOfflineDroppedWhileSiteFailedToStartInRequestHandler_SiteStops_InProcess()
+        {
+            var deploymentResult = await DeployApp(HostingModel.InProcess);
 
-            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot, expectedResponse);
+            // Set file content to empty so it fails at runtime
+            File.WriteAllText(Path.Combine(deploymentResult.DeploymentResult.ContentRoot, "Microsoft.AspNetCore.Server.IIS.dll"), "");
 
-            await AssertAppOffline(deploymentResult, expectedResponse);
+            var result = await deploymentResult.HttpClient.GetAsync("/");
+            Assert.Equal(500, (int)result.StatusCode);
+            Assert.Contains("500.30", await result.Content.ReadAsStringAsync());
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+            AssertStopsProcess(deploymentResult);
```

Comment:
When `app_offline` is added, why is the expected behavior different if the site failed to start in the shim or in the request handler?  Shouldn't the user-visible behavior be the same?

Reply:
If request handler was not loaded it's fine to try again (this is shim failure scenario).
But if inprocess request handler was loaded but was not able to start an application we have to recycle the process because CLR is not restartable.

Reply:
I understand the difference now.  But would it maybe be better to recycle the process when shim fails as well?  Could this prevent customers from manually needing to restart their app, if the shim failed due to some transient error?

Reply:
>  Could this prevent customers from manually needing to restart their app, if the shim failed due to some transient error

I don't really understand how. We would retry when app_offline file would be cycled or configuration changed.